text/0000-cargo-embed-dependency-versions.md

```diff
@@ -0,0 +1,184 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Every time an executable is compiled with Cargo, the dependency tree of the executable is recorded in the binary. This includes the names, versions, dependency kind (build or runtime), and origin (crates.io, git, local filesystem, custom registry). Development dependencies are not recorded, since they cannot affect the final binary. All filesystem paths and URLs are redacted to preserve privacy. The data is encoded in JSON and compressed with zlib to reduce its size.
+
+This data can be recovered using existing tools like `readelf` or Rust-specific tooling. It can be then used to create a Software Bill of Materials in a common format, or audit the dependency list for known vulnerabilities.
+
+WASM, asm.js and embedded platforms are excempt from this mechanism by default since they have very strict code size requirements. For those platforms we encourage you to use tooling that record the hash of every executable in a database and associates the hash with its Cargo.lock, compiler and LLVM version used for the build.
+
+A configuration option can be used to opt out of this behavior if it is not desired (e.g. when building [extremely minimal binaries](https://github.com/johnthagen/min-sized-rust)).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The version information is encoded in an additional arbitrary section of the executable by Cargo. The exact mechanism varies depending on the executable format (ELF, Mach-O, PE, etc.). The section name is subject to bikeshedding.
+
+The data is encoded in JSON which is compressed with Zlib. All arrays a sorted not to disrupt reproducible builds.
+
+The JSON schema specifying the format is provided below. If you find Rust structures more readable, you can find them [here](https://github.com/rust-secure-code/cargo-auditable/blob/311f9932128667b8b18113becdea276b3d98aace/auditable-serde/src/lib.rs#L99-L172). In case of divergences the JSON schema provided in this RFC takes precedence.
+
+```json
+{
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "type": "object",
+    "required": [
+        "packages"
+    ],
+    "properties": {
+        "packages": {
+            "type": "array",
+            "items": {
+                "$ref": "#/definitions/Package"
+            }
+        }
+    },
+    "definitions": {
+        "DependencyKind": {
+            "type": "string",
+            "enum": [
+                "build",
+                "runtime"
```

Comment:
Normally I would consider a "runtime" dependency to be dynamically linked.

In cargo, we call these "normal" dependencies which isn't very clear :)

e.g. from `cargo tree`)
```
  -e, --edges <KINDS>         The kinds of dependencies to display (features, normal, build, dev, all, no-normal,
                              no-build, no-dev, no-proc-macro)
```

Reply:
At this point renaming that causes a format break and requires the whole ecosystem around this (cargo-audit, Syft, Trivy) to transition to these new names. I'm not opposed to this on principle, but we should make sure the changes are worth the breakage.

Reply:
Perhaps you can make `runtime` a deprecated alias of `normal`?

Reply:
That's true, but it breaks less than directly renaming `runtime` => `normal`.

Reply:
Honestly, this is sounding like the "Firefox responds to certain `-webkit-`-prefixed CSS properties" de-facto standardization situation that Rust's system of RFCs and nightlies and `#[feature]` was designed to avoid.

Reply:
> This is still a breaking change. It would require changing all pasers of this data to recognize the `normal` variant.

@Shnatsel Perhaps you can add this to the `auditable` and `audit` implementation to `cargo` nightly only, then add support for it in `cargo-audit`.

Once it stablised, you can then deprecated `cargo-auditable` and `cargo-audit` for latest `cargo` and all the previous `cargo-audit` versions can support decoding embedded information encoded by `cargo`.

Reply:
> Honestly, this is sounding like the "Firefox responds to certain `-webkit-`-prefixed CSS properties" de-facto standardization situation that Rust's system of RFCs and nightlies and `#[feature]` was designed to avoid.

Yeah it's a bit unfortunate.

Reply:
In practice the "runtime" string never appears in the resulting. It is the default case and it is omitted to save space. So the point is rather moot anyway.

Reply:
I've changed it from "runtime" to "normal" for consistency with the help text.